src/event_loop/macos.rs

```diff
@@ -0,0 +1,137 @@
+    /// thread, but if the current thread is the main thread then the task cna also be executed
+    /// directly.
+    executor: Arc<E>,
+
+    /// A background thread for running tasks independently from the host's GUI thread. Useful for
+    /// longer, blocking tasks.
+    background_thread: BackgroundThread<T>,
+
+    /// The reference to the run-loop source so that it can be torn down when this struct is dropped.
+    loop_source: LoopSourceWrapper,
+
+    /// The sender for passing messages to the main thread.
+    main_thread_sender: Sender<T>,
+
+    /// The data that is passed to the external run loop source callback function via a raw pointer.
+    /// The data is not accessed from the Rust side after creating it but it's kept here so as not to get dropped.
+    _callback_data: Box<(Arc<E>, Receiver<T>)>,
+}
+
+impl<T, E> EventLoop<T, E> for MacOSEventLoop<T, E>
+where
+    T: Send + 'static,
+    E: MainThreadExecutor<T> + 'static,
+{
+    fn new_and_spawn(executor: Arc<E>) -> Self {
+        let (main_thread_sender, main_thread_receiver) = channel::bounded::<T>(TASK_QUEUE_CAPACITY);
+
+        let callback_data = Box::new((executor.clone(), main_thread_receiver));
+
+        let loop_source;
+        unsafe {
+            let source_context = CFRunLoopSourceContext {
+                info: &*callback_data as *const _ as *mut c_void,
+                cancel: None,
+                copyDescription: None,
+                equal: None,
+                hash: None,
+                perform: loop_source_callback::<T, E>,
+                release: None,
+                retain: None,
+                schedule: None,
+                version: 0,
+            };
+
+            loop_source = CFRunLoopSourceCreate(
+                kCFAllocatorDefault,
+                1,
+                &source_context as *const _ as *mut CFRunLoopSourceContext,
+            );
+            CFRunLoopAddSource(CFRunLoopGetMain(), loop_source, kCFRunLoopCommonModes);
+        }
+
+        Self {
+            executor: executor.clone(),
+            background_thread: BackgroundThread::new_and_spawn(executor),
+            loop_source: LoopSourceWrapper(loop_source),
+            main_thread_sender,
+            _callback_data: callback_data,
+        }
+    }
+
+    fn schedule_gui(&self, task: T) -> bool {
+        if self.is_main_thread() {
+            self.executor.execute(task, true);
+            true
+        } else {
+            // Only signal the main thread callback to be called if the task was added to the queue.
+            if self.main_thread_sender.try_send(task).is_ok() {
+                unsafe {
+                    CFRunLoopSourceSignal(self.loop_source.0);
+                    CFRunLoopWakeUp(CFRunLoopGetMain());
+                }
+                true
+            } else {
+                false
+            }
+        }
+    }
+
+    fn schedule_background(&self, task: T) -> bool {
+        self.background_thread.schedule(task)
```

Comment:
hmm, just noticed that this call causes memory allocation on mac. more specifically, it seems to be `thread::current().id();` within crossbeam that's causing it. any ideas?

Reply:
now that i think of it, i have a hazy memory of hitting a similar issue with crossbeam in my plugin code which i resolved by switching to [ringbuf](https://github.com/agerasev/ringbuf).

Reply:
Thread locals may allocate when first accessed from a new thread. There's nothing we can do about that. The `thread::current().id()` call just needs to be wrapped in `permit_alloc()`. I'll do that later. The same thing happens in a couple other places if you look through the code base.